### DISCUSSION TITLE
Dashboard implementation in Node-RED

### DISCUSSION THREAD

**Discussion by GrantKeymer** [lbbrhzn/ocpp#326] [Show and tell]
Hello again all,

Over the past few days I've been working away on a Dashboard implementation for my Wallbox Pulsar Plus charger.  After trying to do it with HA automations, I realised how limited they are in their functionality so I've switched to Node-RED.  Below is my first attempt.  Interested to see what you think.
UPDATE - Now with current control slider and timer functions added.

![HA - Node-RED Dashboard](https://user-images.githubusercontent.com/35156198/150905349-98ffb2b6-1f78-4114-8e87-d327fe683071.png)

**Comment by GrantKeymer** [comment]
And here we are, later in the afternoon when the sun has become too weak to continue charging my EV from solar power...

![HA - Node-RED Dashboard 2](https://user-images.githubusercontent.com/35156198/150911738-9d25b32d-9971-4376-8a3e-57921d824072.png)

**Reply by lbbrhzn** [reply]
Looks great! Can you share your code somehow?

**Reply by GrantKeymer** [reply]
Here's what generates the main part of the screen.  In N-R they call it a 'Flow'.  Many of the blocks are self-explanatory but I can share the JS code from the function blocks.  I haven't yet worked out how to share the entire project.

Parts of the screen such as the timer still aren't working and I'm struggling to find out how to connect the 'Set Max Current' slider back into the 'number.wallbox_max_current' entity in HA.  It's relatively easy to read entities from HA into N-R, but not so easy the other way around, especially where it's a variable control, rather than a simple on/off.

UPDATED - To make the flow between some nodes clearer.

![Node-RED Flow 1](https://user-images.githubusercontent.com/35156198/151055976-4bf51254-db22-46c4-bfdc-a33e84430127.png)

**Reply by GrantKeymer** [reply]
Here's what it looks like with the EV battery almost full, during cell re-balancing.  The charge power has fallen to 400 Watts, shortly before the EV says "enough".

![HA - Node-RED Dashboard 4](https://user-images.githubusercontent.com/35156198/151056614-6c8c2ce2-5e29-4cc9-b7f3-2dfa67fcfc00.png)

**Reply by drc38** [reply]
The dashboard looks good, can you stop your charge based on car SoC or range? Your voltage seems rather low... especially if you've got a solar setup?

**Reply by GrantKeymer** [reply]
All the sensors in my Pulsar Plus seem to read between 5 and 10% low compared to what my solar power system says.  For example, the peak charging power reported by my Pulsar Plus is 3.1kW whereas the Redback Solar Power system reports about 3.45kW.  Similarly, the voltage reported by my Pulsar Plus was in the low 220s when I was charging my EV last night whereas the Redback charts showed 241V.  At present, with maximum power of 3.5kW being exported, the Redback reading is 243V.

In common with all AC EV chargers, there is no reporting of SoC through the vehicle connection.  When we come home from a day's driving, I make a mental note of SoC and then work out how many kWh will be needed to get back to 80% or 100%.  If we're driving again the next day, it will sometimes be 100%, depending on how far we're planning to go.  Otherwise, if we're not planning to go anywhere for a while, I leave the EV sitting at around 80%.

For me, it's a very simple equation:  **% SoC gained = kWh x 3.3**.

Where kWh is as indicated by Pulsar Plus, so in reality reads a bit lower than it should do.

**Reply by drc38** [reply]
If your vehicle is supported one of these can be useful (fun to tinker with) to get metrics on the car/battery https://www.openvehicles.com/

**Reply by GrantKeymer** [reply]
Thanks for the tip.  I see there is a cable specifically for the Nissan Leaf, so I can't imagine it wouldn't be supported, as the first mass-produced EV.  Another idea to investigate when I have the time 😄

**Reply by drc38** [reply]
Yes it is. I've extended the ovms code to support my 2016 AZE0 and V2H, and another NZer contributed to the ZE0 support.

**Reply by GrantKeymer** [reply]
Do you have a V2H inverter installed yet?

**Comment by GrantKeymer** [comment]
Now I have some additional nodes working...

![Node-RED Flow 2](https://user-images.githubusercontent.com/35156198/151106183-3ec834ea-1846-4f18-826c-75a3d79ecd35.png)

Meaning that the 'Set Max Current' slider in my dashboard communicates all the way through to the Wallbox App and charger.  It was a lot harder than it should have been due to poor documentation by the Node-RED team, but I got there in the end 😁

**Comment by stamateviorel** [comment]
can you share your configuration ?

**Reply by sincze** [reply]
Yes would be nice to know how to set this up in NodeRed directly without having the need for HA.
You can export a flow quite easily
https://nodered.org/docs/user-guide/editor/workspace/import-export

**Comment by dzwigar** [comment]
great job. can you share your flow?

**Reply by longobardi-davide** [reply]
It's a great job, could you share it or public the flow? If that I would like to try to add mqtt to it.

**Comment by lwefq** [comment]
Wow this looks great! I'm also thinking of using Node-RED in order to implement the dashboard for my charger. Could you share the Node-RED flow? Thanks!

**Comment by GrantKeymer** [comment]
This is a pretty old thread so my Dashboard looks very different now...

![EV Charger Dashboard Dec 23](https://github.com/lbbrhzn/ocpp/assets/35156198/3cf6d25c-30bb-4147-923c-b55eb7a9849d)

I stopped using OCPP a long time ago and have been using MQTT to control my Wallbox Pulsar Plus for about 3 months now.  I've made a version of the Flow with all personal data removed, so if you'd like a copy, please send me a PM.

**Reply by sincze** [reply]
> This is a pretty old thread so my Dashboard looks very different now...
> 
> ![EV Charger Dashboard Dec 23](https://github.com/lbbrhzn/ocpp/assets/35156198/3cf6d25c-30bb-4147-923c-b55eb7a9849d)
> 
> I stopped using OCPP a long time ago and have been using MQTT to control my Wallbox Pulsar Plus for about 3 months now.  I've made a version of the Flow with all personal data removed, so if you'd like a copy, please send me a PM.

PM ;)

I was looking for that however i found:

You can't send a direct private message to another GitHub user. Alternative ways to communicate is through their open source projects. If they have one, you could open an issue and ask a question related to that project.

Hence i try it this way could you consider this a PM :)

**Reply by GrantKeymer** [reply]
@sincze - OK, I'll consider it a PM 🙂

Meanwhile, please email me at...

grant.keymer [at] gmail.com 

Replace [at] with the magic symbol 😉

**Comment by dzwigar** [comment]
Hi.
Is it possible to communicate with Wallbox pulsar with nodered? I'm asking you for help.

**Reply by rsteininger** [reply]
hi @dzwigar,
I had done that based on a python script I had found and it works relatively stable. 
Pls give me a day or two and I will share my flow

**Reply by rsteininger** [reply]
@dzwigar 
This is my flow I use in node-red for my wallbox pulsar. Only needs your user name / pw and if you want the ip address to monitor if the charger is online.
[wallbox-pulsar.json](https://github.com/user-attachments/files/16553186/wallbox-pulsar.json)

**Reply by jjotto** [reply]
@rsteininger where can I find the missing nodes?  There may be an easy way, but I'm not seeing it.3
![image](https://github.com/user-attachments/assets/dd5f9c1e-e38c-4fa9-8484-3247fccb556e)

**Reply by jjotto** [reply]
self reply!  https://flows.nodered.org/inspect   Flow Inspector!
